src/Microsoft.AspNetCore.Mvc.Core/Internal/DefaultAssemblyPartDiscoveryProvider.cs

```diff
@@ -63,14 +63,100 @@ internal static IEnumerable<RuntimeLibrary> GetCandidateLibraries(DependencyCont
                 return Enumerable.Empty<RuntimeLibrary>();
             }
 
-            return dependencyContext.RuntimeLibraries.Where(IsCandidateLibrary);
+            var candidatesResolver = new CandidatesResolver(dependencyContext.RuntimeLibraries, ReferenceAssemblies);
+            return candidatesResolver.GetCandidates();
         }
 
-        private static bool IsCandidateLibrary(RuntimeLibrary library)
+        private class CandidatesResolver
         {
-            Debug.Assert(ReferenceAssemblies != null);
-            return !ReferenceAssemblies.Contains(library.Name) &&
-                library.Dependencies.Any(dependency => ReferenceAssemblies.Contains(dependency.Name));
+            private readonly IDictionary<string, Dependency> _dependencies;
+
+            public CandidatesResolver(IReadOnlyList<RuntimeLibrary> dependencies, ISet<string> referenceAssemblies)
+            {
+                _dependencies = dependencies
```

Comment:
Is the sequence of `RuntimeLibrary` guaranteed to be unique? `ToDictionary` will throw if two or items have the same key.